src/context_free.ml

```diff
@@ -8,6 +8,34 @@ module A = Attribute
 module AC = Attribute.Context
 
 module Rule = struct
+  module Attribute_list = struct
```

Comment:
This and the `Parsed_payload_list` seem extremly convoluted to me and I'm not sure they bring a lot of value.

I might be wrong but it seems the same thing can be achieved using a single attribute and passing all the config in the payload, e.g. instead of:
```ocaml
let x = y [@prefix "pre"][@suffix "post]
```
one could write:
```ocaml
let x = y [@fix ~pre:"pre" ~suf:"post"]
```
or something along those lines.

It would also clarify things greatly from a reader perspective, as it's now obvious this is a single transformation handling both, the name of the attribute likely hinting to the name of the ppx responsible for handling it.

Reply:
What's going on here, is that we have multiple attributes that are often used separately, but may be used together. We want each one to be succinct. When transformed as just a collection of attributes, they need to be combined to get the right result. But we don't want
```
x [@foo a], y [@bar b], z [@foo a] [@bar b]
```
to have to be written as
```
x [@foobar ~foo:a], y [@foobar ~bar:b], z [@foobar ~foo:a ~bar:b]
```
That's much more verbose. The fact that these things are transformed together is sort of an implementation detail, and I don't think it needs to creep into the notation.

Reply:
I think this is still achievable with the suggested API. One could register rules for the individual attributes for conciseness and have a main attribute that lets you combine them.

As you mention, the fact that you have to handle them together is an implementation detail, it probably shouldn't leak into ppxlib public API.

Would something like the following example be acceptable?

E.g. you'd write:
```ocaml
x [@foo a], y [@bar b], z [@foobar ~foo:a ~bar:b]
```

This would also likely make the ppx code a bit simpler as you don't have to handle the GADT lists in there anymore either.

Reply:
The issue is that people would use those shorter attributes together, which means that the ppx code would need to handle having multiple attributes that it cares about on one node. every ppx will have to write code that either tries to consume all of its other attributes or at least spots them and emits an error. The idea behind this code was to move that code into ppxlib so that the individual ppxs don't have to worry about it.

I agree the GADT lists are a tad complicated to look at, however they're surprisingly easy to use, which is what convinced me that this is probably fine, because they define `[]` and `( :: )` you can basically just pretend they're normal lists in the ppx code, but just get a compile error if it's the wrong length or has a wrong element type.

Reply:
Perhaps we should (1) put the single-attribute case first and document that, then (2) present the multiple-attribute case in a submodule. The submodule's documentation can describe a use case and show how it works, so by the time the user sees the GADTs and the function signature, it's much clearer what it's all for.

Reply:
@NathanReb does that sound sufficient to you?

Reply:
@ceastlund Yes that sounds good to me. @Skepfyr made a great argument there as well and I can see how neat it is to use.

@patricoferris also pointed out that we have an `Expert` module in `Extension`. We could do something similar here.

I'm also wondering if we shouldn't move parts of this to its own file as it describes a specific type of rules, just as extensions do.

Reply:
I've extracted it out into a submodule. I don't think creating an entirely separate file like extension or deriving is quite worth it because there's a lot less interface, but happy to move it if you think otherwise.

Reply:
Looks good to me, thanks!